HLTrigger/special/plugins/HLTPPSCalFilter.cc

```diff
@@ -0,0 +1,183 @@
+
+private:
+  edm::InputTag pixelLocalTrackInputTag_;  // Input tag identifying the pixel detector
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSPixelLocalTrack>> pixelLocalTrackToken_;
+
+  edm::InputTag stripLocalTrackInputTag_;  // Input tag identifying the strip detector
+  edm::EDGetTokenT<edm::DetSetVector<TotemRPLocalTrack>> stripLocalTrackToken_;
+
+  edm::InputTag diamondLocalTrackInputTag_;  // Input tag identifying the diamond detector
+  edm::EDGetTokenT<edm::DetSetVector<CTPPSDiamondLocalTrack>> diamondLocalTrackToken_;
+
+  int minTracks_;
+  int maxTracks_;
+
+  bool do_express_;
+};
+
+void HLTPPSCalFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("pixelLocalTrackInputTag", edm::InputTag("ctppsPixelLocalTracks"))
+      ->setComment("input tag of the pixel local track collection");
+  desc.add<edm::InputTag>("stripLocalTrackInputTag", edm::InputTag("totemRPLocalTrackFitter"))
+      ->setComment("input tag of the strip local track collection");
+  desc.add<edm::InputTag>("diamondLocalTrackInputTag", edm::InputTag("ctppsDiamondLocalTracks"))
+      ->setComment("input tag of the diamond local track collection");
+
+  desc.add<int>("minTracks", 1)->setComment("minimum number of tracks in pot");
+  desc.add<int>("maxTracks", -1)->setComment("maximum number of tracks in pot");
+
+  desc.add<bool>("do_express", true)->setComment("toggle on filter type; true for Express, false for Prompt");
+
+  desc.add<int>("triggerType", trigger::TriggerTrack);
+
+  descriptions.add("hltPPSCalFilter", desc);
+}
+
+HLTPPSCalFilter::HLTPPSCalFilter(const edm::ParameterSet& iConfig)
+    : pixelLocalTrackInputTag_(iConfig.getParameter<edm::InputTag>("pixelLocalTrackInputTag")),
+      pixelLocalTrackToken_(consumes<edm::DetSetVector<CTPPSPixelLocalTrack>>(pixelLocalTrackInputTag_)),
+
+      stripLocalTrackInputTag_(iConfig.getParameter<edm::InputTag>("stripLocalTrackInputTag")),
+      stripLocalTrackToken_(consumes<edm::DetSetVector<TotemRPLocalTrack>>(stripLocalTrackInputTag_)),
+
+      diamondLocalTrackInputTag_(iConfig.getParameter<edm::InputTag>("diamondLocalTrackInputTag")),
+      diamondLocalTrackToken_(consumes<edm::DetSetVector<CTPPSDiamondLocalTrack>>(diamondLocalTrackInputTag_)),
+
+      minTracks_(iConfig.getParameter<int>("minTracks")),
+      maxTracks_(iConfig.getParameter<int>("maxTracks")),
+
+      do_express_(iConfig.getParameter<bool>("do_express")) {}
+
+bool HLTPPSCalFilter::filter(edm::StreamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+  // Helper tool to count valid tracks
+  const auto valid_trks = [](const auto& trk) { return trk.isValid(); };
+
+  // maps for assigning filter pass / fail
+  std::unordered_map<uint32_t, bool> pixel_filter_;
+  std::unordered_map<uint32_t, bool> strip_filter_;
+  std::unordered_map<uint32_t, bool> diam_filter_;
```

Comment:
It seems these 3 maps are just used as bools to switch on/off IDs, so these variables could become (const) class members and initialised as such in the class, rather than being local variables.

Reply:
The maps are used as bools to store whether the track multiplicity of each detector is within the max and min number of tracks. The final filter condition combines the result for each detector in a single logical expression, depending on which type of filtering is done. 
Given their purpose, I'm not sure how to define the maps as const class members and initialize them in the class, and still be able to change their value when the filter checks track multiplicity conditions. Could you give an example of how to implement this?